pzsd_bot/__main__.py

```diff
@@ -39,10 +54,132 @@ async def on_message(message):
+            logger.info(
+                "%s attempted to give themselves %s points. Very naughty.",
+                bestower.name,
+                point_amount,
+            )
+        else:
+            logger.info(
+                "%s awarding %s point(s) to %s",
+                bestower.name,
+                point_amount,
+                recipient.name,
+            )
+
+        if valid_transaction:
+            async with engine.begin() as conn:
+                await conn.execute(
+                    insert(ledger).values(
+                        bestower=bestower.id,
+                        recipient=recipient.id,
+                        points=point_amount,
+                    )
+                )
+                logger.info("Added point transaction to ledger")
+
+        if valid_transaction:
+            title = "Point transaction"
+            color = 0xFFFFFF
+        else:
+            title = "Self point violation!"
+            color = 0xFF0000
+        embed = discord.Embed(
+            title=title,
+            description=f"[Jump to original message]({message.jump_url})",
+            colour=color,
+            timestamp=datetime.now(),
+        )
+        embed.add_field(name="Bestower", value=bestower.name, inline=True)
+        embed.add_field(name="Recipient", value=recipient.name, inline=True)
+        embed.add_field(name="Point amount", value=str(point_amount), inline=True)
+        message_content = message.content
+        if len(message_content) > 80:
+            message_content = message_content[:80] + "..."
+        embed.add_field(name="Content of message:", value=message_content, inline=False)
+
+        points_log_channel = bot.get_channel(int(os.environ["POINTS_LOG_CHANNEL"]))
+        await points_log_channel.send(embed=embed)
+
 
 @bot.slash_command(guild_ids=[GUILD_ID])
-async def test(ctx):
-    await ctx.respond("This was a test.")
+async def leaderboard(ctx):
+    points = {}
+
+    async with engine.connect() as conn:
+        result = await conn.execute(select(pzsd_user))
+        for row in result.fetchall():
+            points[row.id] = {"name": row.name, "point_total": 0}
+
+        result = await conn.execute(select(ledger))
+        for row in result:
+            points[row.recipient]["point_total"] += row.points
+
+    embed = discord.Embed(title="Points Leaderboard", colour=0xA8A434)
+    for i, user in enumerate(
+        sorted(points.values(), key=lambda d: d["point_total"], reverse=True), 1
+    ):
+        name = user["name"].capitalize()
+        point_total = user["point_total"]
+        embed.add_field(
+            name=f"{i}. {name}", value=f"{point_total:,} points", inline=False
+        )
+
+    await ctx.respond(embed=embed)
+
+
+async def run_bot():
+    try:
+        await bot.start(BOT_TOKEN)
+    finally:
+        await engine.dispose()
 
 
-bot.run(BOT_TOKEN)
+try:
+    asyncio.run(run_bot())
+except KeyboardInterrupt:
+    print("Exiting...")
```

Comment:
use logger

Reply:
This I have to push back on, it doesn't make sense to log this, especially since each time the bot is started up, the log is overwritten. So given this only appears once when halting the bot, it's not useful to log. I just print it to have a clean exit message seen in stdout when sending a keyboard interrupt.

Reply:
Now that I think about it, idk why i'm even truncating the log file every time, I'll probably change it to append. But still, I don't really see why I would log this.

Reply:
I changed my mind